### PR TITLE
Improve bounced messages implementation (NNBB)

### PR DESCRIPTION
**Pull request by nuno-vieira**
### 🔗 Issue Links
Resolves https://github.com/GetStream/ios-issues-tracking/issues/563

### 🎯 Goal
This PR aims to improve the NNBB feature added previously here: https://github.com/GetStream/stream-chat-swift/pull/2103. That PR was the initial implementation and more of a PoC (AFAIK). The feature has been completely changed since then from the backend, which had breaking changes.

**Note:** This PR includes some minor breaking changes. This is fine because the feature was not working either way and was just a PoC.

### 📝 Summary
- Adds message moderation details to `ChatMessage`
- Fixes bounced message displayed as a system message instead of an error
- Add support for resending bounced messages
- Add support for deleting bounced messages
- Fix not being able to delete local-only messages
- Fix not showing bounced actions when long pressing bounced message
- Move responsibility of showing the bounced alert actions to the message list instead of Message Actions.

### 🎨 Showcase
https://github.com/GetStream/stream-chat-swift/assets/12814114/559f0809-5957-437f-a874-a7298d5b73a0

#### Precondition
Go to the moderation dashboard [here](https://moderation.getstream.io/app/102399/moderation/settings/channel-type/messaging). To change the moderation action, go to **Commercial Spam** > **Configure** > **Select action**.

**Scenario 1 - Infinite Bounce:**
1. In the Moderation Dashboard, select **Commercial Spam** > **Configure** > **Bounce**
2. Go to a Channel and send a spam message like this: "Click here to win a free iPhone!!!!"
3. The message should have an error indicator
4. When long pressing or taping the error indicator, it should appear an alert with actions
       a. Send Anyway -> The message should be bounced again
       b. Edit Message -> Edit the message until it is valid and not bounced
       c. Delete Message -> Hard deletes the message

**Scenario 2 - Flag:**
1. In the Moderation Dashboard, select **Commercial Spam** > **Configure** > **Flag**
2. Go to a Channel and send a spam message like this: "Click here to win a free iPhone!!!!"
3. The message should have the regular UI
4. The message should appear in the Moderation Dashboard [here](https://moderation.getstream.io/app/102399/moderation/messages).

**Scenario 3 - Block:**
1. In the Moderation Dashboard, select **Commercial Spam** > **Configure** > **Block**
2. Go to a Channel and send a spam message like this: "Click here to win a free iPhone!!!!"
3. The message should be rendered as a system message like so: "This message has been blocked by Moderation", and it is only shown on the sender's device.
4. The message should appear in the Moderation Dashboard [here](https://moderation.getstream.io/app/102399/moderation/messages).

**Scenario 4 - Bounce then Flag:**
1. In the Moderation Dashboard, select **Commercial Spam** > **Configure** > **Bounce then Flag**
2. Go to a Channel and send a spam message like this: "Click here to win a free iPhone!!!!"
3. The message should have an error indicator
4. When long pressing or taping the error indicator, it should appear an alert with actions
       a. Send Anyway -> The message should be flagged (Scenario 2)
       b. Edit Message -> Editing to a valid message it will have the regular flow, if not, it is flagged (Scenario 2).
       c. Delete Message -> Hard deletes the message

**Scenario 5 - Bounce then Block:**
1. In the Moderation Dashboard, select **Commercial Spam** > **Configure** > **Bounce then Block**
2. Go to a Channel and send a spam message like this: "Click here to win a free iPhone!!!!"
3. The message should have an error indicator
4. When long pressing or taping the error indicator, it should appear an alert with actions
       a. Send Anyway -> The message should be blocked (Scenario 3)
       b. Edit Message -> Editing to a valid message will have the regular flow, if not, it is blocked (Scenario 3).
       c. Delete Message -> Hard deletes the message

### ☑️ Contributor Checklist

- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] This change follows zero ⚠️ policy (required)
- [x] This change should be manually QAed
- [x] Changelog is updated with client-facing changes
- [x] New code is covered by unit tests
- [x] Comparison screenshots added for visual changes
- [ ] Affected documentation updated (docusaurus, tutorial, CMS)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -3,6 +3,15 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 
 # Upcoming
 
+## StreamChat
+### ✅ Added
+- Add message moderation details to `ChatMessage` [#2846](https://github.com/GetStream/stream-chat-swift/pull/2846)
+- Add support for resending bounced messages [#2846](https://github.com/GetStream/stream-chat-swift/pull/2846)
```
Comment (martinmitrevski) [inline CHANGELOG.md:9]: maybe merge these 3 in one?

File: CHANGELOG.md [hunk 2]
```diff
@@ -26,6 +35,11 @@ _October 25, 2023_
     - Enabled by `Components.default.messageAutoTranslationEnabled`
 ### 🐞 Fixed
 - Do not mark a channel as read if the app is in background [#2832](https://github.com/GetStream/stream-chat-swift/pull/2832)
```
Comment (martinmitrevski) [inline CHANGELOG.md:37]: Hm, are you sure you correctly merged the changelog? This is from the previous release.

File: Sources/StreamChat/Workers/MessageUpdater.swift [hunk 3]
```diff
@@ -63,14 +63,10 @@ class MessageUpdater: Worker {
 
             // Hard Deleting is necessary for bounced messages, since these messages are never stored on the cloud
             // an apiClient request to delete them would never be triggered.
-            let shouldBeHardDeleted = hard || messageDTO.failedToBeSentDueToModeration
-            let shouldAllowLocallyStoredMessagesToBeDeleted = !isLocalStorageEnabled
```
Comment (martinmitrevski) [inline Sources/StreamChat/Workers/MessageUpdater.swift:67]: We aren't missing a case from here, right? An explanation would be nice.

File: Sources/StreamChatUI/MessageActionsPopup/ChatMessageActionsVC.swift [hunk 4]
```diff
@@ -83,15 +83,6 @@ open class ChatMessageActionsVC: _ViewController, ThemeProvider {
         }
     }
 
-    open var messageActionsForAlertMenu: [ChatMessageActionItem] {
```
Comment (martinmitrevski) [inline Sources/StreamChatUI/MessageActionsPopup/ChatMessageActionsVC.swift:86]: Why is this one removed?

**Review by martinmitrevski — COMMENTED** [review]
Looks good 👍 I added few small notes and questions

File: Sources/StreamChatUI/Utils/ChatMessage+Extensions.swift [hunk 5]
```diff
@@ -8,17 +8,26 @@ import StreamChat
...
     var isInteractionEnabled: Bool {
-        guard
-            type != .ephemeral, type != .system, type != .error,
-            isDeleted == false
-        else { return false }
+        if isDeleted {
+            return false
+        }
+
+        if shouldRenderAsSystemMessage {
```
Comment (polqf) [inline Sources/StreamChatUI/Utils/ChatMessage+Extensions.swift:15]:
```suggestion
        if isDeleted || shouldRenderAsSystemMessage {
```

File: CHANGELOG.md (hunk 1 above)
Comment (nuno-vieira) [reply CHANGELOG.md]: Done ✅

File: CHANGELOG.md (hunk 2 above)
Comment (nuno-vieira) [reply CHANGELOG.md:37]: Fixed ✅

File: Sources/StreamChat/Models/MessageModerationDetails.swift [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+/// The type of moderation performed to a message.
+public struct MessageModerationAction: RawRepresentable, Equatable {
+    public var rawValue: String
+
+    public init(rawValue: String) {
+        self.rawValue = rawValue
+    }
+
+    /// A bounced message means it needs to be rephrased and sent again.
+    static var bounce = Self(rawValue: "MESSAGE_RESPONSE_ACTION_BOUNCE")
```
[the remark this answers is not in the document]
Comment (nuno-vieira) [reply Sources/StreamChat/Models/MessageModerationDetails.swift:24]: Nice catch 👍 Fixed ✅

File: Sources/StreamChat/Workers/MessageUpdater.swift (hunk 3 above)
Comment (nuno-vieira) [reply Sources/StreamChat/Workers/MessageUpdater.swift:67]: No, the previous logic was super confusing. Right now it is pretty simple, and the tests should be self-explanatory as well as the code 🤔 We should hard delete a message if `hard` is equal to `true` or if the message is local only.

File: Sources/StreamChatUI/MessageActionsPopup/ChatMessageActionsVC.swift (hunk 4 above)
Comment (nuno-vieira) [reply Sources/StreamChatUI/MessageActionsPopup/ChatMessageActionsVC.swift:86]: It did not make sense for the `ChatMessageActionsVC` to provide the actions for the AlertController in the `ChatMessageListVC`. The `ChatMessageActionsVC` is used for long pressing a normal message, and it shows the Popup view. When a message is bounced, it shows an alert controller which has nothing to do with the `ChatMessageActionsVC`. So this property was moved to the Message List with the name of `messageActions(forBouncedMessage:) -> [ChatMessageActionItem]`

File: Sources/StreamChat/Workers/MessageUpdater.swift (hunk 3 above)
Comment (nuno-vieira) [reply Sources/StreamChat/Workers/MessageUpdater.swift:67]: I will update the comment tho 👍

File: Sources/StreamChatUI/MessageActionsPopup/ChatMessageActionsVC.swift (hunk 4 above)
Comment (martinmitrevski) [reply Sources/StreamChatUI/MessageActionsPopup/ChatMessageActionsVC.swift:86]: If this was used only for bounced messages, then it's ok (the method's name didn't clearly state that).

**Review by martinmitrevski — APPROVED** [review]

File: Sources/StreamChatUI/MessageActionsPopup/ChatMessageActionsVC.swift (hunk 4 above)
Comment (nuno-vieira) [reply Sources/StreamChatUI/MessageActionsPopup/ChatMessageActionsVC.swift:86]: Yeah it was used only for bounced messages. The previous name was indeed not clear.

**Pull request merged**